packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -557,162 +558,118 @@ impl<'m> RunnerImpl<'m> {
-            DataType::Float64 => {
-                buffer_lens.push(target_len * 8); // 8 bytes per f64
-                Ok(())
-            }
-            DataType::UInt32 => {
-                buffer_lens.push(target_len * 4); // 4 bytes per u32
-                Ok(())
-            }
-            DataType::UInt16 => {
-                buffer_lens.push(target_len * 2); // 2 bytes per u16
-                Ok(())
-            }
-            DataType::Utf8 => {
-                // TODO: Use `data.len` or target_len?
-                //       (In practice, target_len has worked for a long time,
-                //       though that's not an ideal reason to use it. Maybe
-                //       `data.len` would also work.)
-                let offsets = unsafe {
-                    std::slice::from_raw_parts(data.buffer_ptrs[0] as *const i32, target_len + 1)
-                };
-                debug_assert_eq!(offsets[0], 0);
-                let last = offsets[target_len];
-                // offsets
-                buffer_lens.push((target_len + 1) * 4);
-                buffer_lens.push(last as usize);
-                Ok(())
-            }
-            DataType::List(_) => {
-                // offsets
-                buffer_lens.push((target_len + 1) * 4);
-                Ok(())
-            } // Just offsets
-            DataType::Struct(_) => Ok(()), // No non-child buffers
-            DataType::FixedSizeList(..) => Ok(()),
-            DataType::FixedSizeBinary(sz) => {
-                buffer_lens.push(data.len * sz as usize);
-                Ok(())
-            } // Just values
-            DataType::Boolean => {
-                buffer_lens.push((data.len / 8) + (if data.len % 8 == 0 { 0 } else { 1 }));
-                Ok(())
-            } // Just values
-            t => Err(Error::FlushType(t)), // TODO: More types?
-        }?;
-
-        debug_assert_eq!(data.n_buffers, buffer_lens.len());
-        let mut buffers = Vec::new();
-        for (i, &len) in buffer_lens.iter().enumerate().take(data.n_buffers) {
-            let ptr = data.buffer_ptrs[i];
-            debug_assert_ne!(ptr, std::ptr::null());
-            let capacity = data.buffer_capacities[i];
-            let buffer = if len <= capacity {
-                unsafe { self.new_buffer(ptr, len, capacity) }
-            } else {
-                // This happens when we have fixed size buffers, but the inner nodes are null
-                let mut mut_buffer = MutableBuffer::new(len);
-                mut_buffer.resize(len, 0);
-                mut_buffer.into()
-            };
-            // let buffer = unsafe { Buffer::from_unowned(ptr, len, capacity) };
-            buffers.push(buffer);
-        }
+        builder = builder.len(target_len);
+        if !data.null_bits_ptr.is_null() {
+            let mut boolean_builder = BooleanBufferBuilder::new(target_len);
+            let values = unsafe { slice::from_raw_parts(data.null_bits_ptr, target_len) };
```

Comment:
Inconsistent with the unsafe blocks above in terms of how much is inside its scope